### DISCUSSION TITLE
how to watch void* type actual value?

### DISCUSSION THREAD

**Discussion by pily1** [OpenModelica/OpenModelica#12414] [Q&A] [answered]
I debug omc.exe with vscode,I try to convert a mo-file to C file,I want to know how the mo file convert to a C file,
so I add breadpoint int the CodegenC.c  file omc_CodegenC_simulationFunctionsFile function,
I add a watch to watch _out_txt value,but _out_txt is void* type,is a address,
how can I watch the actual value?

**Comment by casella** [comment, marked as answer]
@pily1  that is a very, very, very long story 😃.

You can find an overview of the structure of the compiler in [this presentation](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://openmodelica.org/images/docs/openmodelica2015/OpenModelica2015-talk08-adrpo-OpenModelica.pdf&ved=2ahUKEwi4vufdho2GAxUU0wIHHaeoDTcQFnoECA8QAQ&usg=AOvVaw30JpRcH5ZzHx1ag3B2nNp5). Maybe also have a look at [this paper](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://www.mic-journal.no/PDF/2020/MIC-2020-4-1.pdf&ved=2ahUKEwjR6v-Nh42GAxXuxQIHHao6CCYQFnoECBAQAQ&usg=AOvVaw27wB4Kl6TPidzUpsnSHTel).

**Reply by pily1** [reply]
think you,you have mentioned the paper is very useful.

**Comment by AnHeuermann** [comment]
Check the functions defined in https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/meta/meta_modelica.c. As far as I know they are used for the debugger interface to show what type a variable is and so on.
Maybe try `printTypeOfAny` and see what it reports. It's somewhere on my TODO list to understand this part, but I haven't started yet.
You can execute the C functions, they are part of omc.exe. Run them from GDB for some memory address and see what they print. But the could also crash.

And then there are the MDT debugging plugin for Eclipse (as @casella mentioned) and the algorithm debugger from OMEdit which can do something like this as well. But I don't know how they are working.

**Reply by pily1** [reply]
think you ,I will try it
